sw/device/lib/runtime/hmac.c

```diff
@@ -0,0 +1,84 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "sw/device/lib/runtime/hmac.h"
+
+#include "sw/device/lib/base/bitfield.h"
+#include "sw/device/lib/base/mmio.h"
+
+#include "hmac_regs.h"  // Generated.
+#include "hw/top_earlgrey/sw/autogen/top_earlgrey.h"
+
+void hmac_sha256_init(void) {
+  mmio_region_t hmac_base = mmio_region_from_addr(TOP_EARLGREY_HMAC_BASE_ADDR);
+
+  // Clear the config, stopping the SHA engine.
+  mmio_region_write32(hmac_base, HMAC_CFG_REG_OFFSET, 0u);
+
+  // Disable and clear interrupts. INTR_STATE register is rw1c.
+  mmio_region_write32(hmac_base, HMAC_INTR_ENABLE_REG_OFFSET, 0u);
+  mmio_region_write32(hmac_base, HMAC_INTR_STATE_REG_OFFSET, 0u);
+
+  uint32_t reg = 0;
+  // Digest is little-endian by default.
+  reg = bitfield_bit32_write(reg, HMAC_CFG_DIGEST_SWAP_BIT, false);
+  // Enable endian swap since our inputs are little-endian.
+  reg = bitfield_bit32_write(reg, HMAC_CFG_ENDIAN_SWAP_BIT, true);
+  reg = bitfield_bit32_write(reg, HMAC_CFG_SHA_EN_BIT, true);
+  reg = bitfield_bit32_write(reg, HMAC_CFG_HMAC_EN_BIT, false);
+  mmio_region_write32(hmac_base, HMAC_CFG_REG_OFFSET, reg);
+
+  reg = 0;
+  reg = bitfield_bit32_write(reg, HMAC_CMD_HASH_START_BIT, true);
+  mmio_region_write32(hmac_base, HMAC_CMD_REG_OFFSET, reg);
+}
+
+hmac_error_t hmac_sha256_update(const void *data, size_t len) {
+  if (data == NULL) {
+    return kHmacErrorBadArg;
+  }
+  mmio_region_t hmac_base = mmio_region_from_addr(TOP_EARLGREY_HMAC_BASE_ADDR);
+  const uint8_t *data_sent = (const uint8_t *)data;
+
+  // Individual byte writes are needed if the buffer isn't word aligned.
+  for (; len != 0 && (uintptr_t)data_sent & 3; --len) {
```

Comment:
There is a function named `misalign` or similar in memory.h that may be helpful for doing this a bit more cleanly.

Reply:
There's `misalignment32_of` but it's not exposed by the header. Should I expose it?

Reply:
Hmm, maybe. I might be thinking of another utility. Also ok leaving this as is... @alphan?

Reply:
We have `misalignment_of()` in `mmio.c`. We should probably move it to `memory`. I'm ok with leaving this as is until we hit other cases.